components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.SynapseEnvironment;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.Authenticator;
+import org.wso2.carbon.apimgt.gateway.handlers.security.oauth.OAuthAuthenticator;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+/**
+ * Authenticator responsible for handling Mutual SSL and OAuth secured APIs.
+ */
+public class MutualSSLAndOAuthAuthenticator implements Authenticator {
+    private static final Log log = LogFactory.getLog(MutualSSLAndOAuthAuthenticator.class);
+    private volatile Authenticator oAuthAuthenticator;
+    private volatile Authenticator mutualSSLAuthenticator;
+    private String apiLevelPolicy;
+    private String authorizationHeader;
+    private String certificateInformation;
+    private boolean removeOAuthHeadersFromOutMessage;
+
+    /**
+     * Initialized the authenticator with required parameter.
+     *
+     * @param authorizationHeader    Relevant authorization header.
+     * @param removeOAuthHeader      indicate whether to remove OAuth header.
+     * @param apiLevelPolicy         API level tier policy
+     * @param certificateInformation Relevant certificate information.
+     */
+    public MutualSSLAndOAuthAuthenticator(String authorizationHeader, boolean removeOAuthHeader,
+            String apiLevelPolicy, String certificateInformation) {
+        this.authorizationHeader = authorizationHeader;
+        this.removeOAuthHeadersFromOutMessage = removeOAuthHeader;
+        this.apiLevelPolicy = apiLevelPolicy;
+        this.certificateInformation = certificateInformation;
+    }
+
+    @Override
+    public void init(SynapseEnvironment env) {
+        getAuthenticator(APIConstants.DEFAULT_API_SECURITY_OAUTH2).init(env);
+        getAuthenticator(APIConstants.API_SECURITY_MUTUAL_SSL).init(env);
+    }
+
+    /**
+     * To get the relevant authenticator with the authenticator name.
+     *
+     * @param authenticatorName Name of authenticator.
+     * @return relevant authenticator name.
+     */
+    private Authenticator getAuthenticator(String authenticatorName) {
+        Authenticator authenticator;
+        if (authenticatorName.equalsIgnoreCase(APIConstants.DEFAULT_API_SECURITY_OAUTH2)) {
+            if (oAuthAuthenticator == null) {
+                oAuthAuthenticator = new OAuthAuthenticator(authorizationHeader, removeOAuthHeadersFromOutMessage);
+            }
+            authenticator = oAuthAuthenticator;
+        } else {
+            if (mutualSSLAuthenticator == null) {
+                mutualSSLAuthenticator = new MutualSSLAuthenticator(apiLevelPolicy, certificateInformation);
+            }
+            authenticator = mutualSSLAuthenticator;
+        }
+        return authenticator;
+    }
```

Comment:
I think this if-else branch has no use. why can't we simply instantiate and init 2 authenticators at line 60-61?